allennlp/data/token_indexers/pretrained_transformer_indexer.py

```diff
@@ -65,14 +87,24 @@ def count_vocab_items(self, token: Token, counter: Dict[str, Dict[str, int]]):
         pass
 
     @overrides
-    def tokens_to_indices(self, tokens: List[Token], vocabulary: Vocabulary) -> IndexedTokenList:
+    def tokens_to_indices(
+        self, tokens: List[Union[Token, int]], vocabulary: Vocabulary
+    ) -> IndexedTokenList:
+        """
+        `tokens` may already be indices, in which case the token -> index step is a no-op, but other
```

Comment:
When will `tokens` already be indices?

Reply:
This is basically making the mismatched case a little easier. With some changes introduced in this PR, now the mismatched indexer has to do subword tokenization AND special token insertion before [calling](https://github.com/allenai/allennlp/blob/master/allennlp/data/token_indexers/pretrained_transformer_mismatched_indexer.py#L64) the matched, i.e. default, indexer's `tokens_to_indices()`. Because we use huggingface's `_build_input_with_special_tokens(List[int]) -> List[int]` method to insert special tokens, we already have a list of `int` indices before calling the matched indexer's `tokens_to_indices()`. Technically we can convert back to `Token`s and then leave this place unchanged, but I think it's just a little easier this way.

Reply:
In that case you should use a `# type: ignore` on the caller, not change the type annotation here.  What you did here should actually make mypy complain, because now the type signature doesn't match the superclass.

Reply:
I'm not sure why mypy isn't complaining, actually.  Maybe it was that pylint would have complained, but we got rid of pylint?

Reply:
I _am_ using a `# type: ignore` on the caller -- you can search for the string `)  # type: ignore` on the `Files Changed` page.

Yeah I didn't realize the mismatch with superclass issue. Are you suggesting to leave the type signature here unchanged (i.e. back to `List[Token]`) but actually expects that `tokens` might be `List[int]` in the function body?

Reply:
Ideally though I think we should break apart the `Token -> Index` logic and any potential postprocessing logic (e.g. the long sequence splitting one here) into two separate functions. And if we already have indices, we can skip the former function and directly call the second. But here we only have this one function so it is a little weird that a function called `tokens_to_indices()` actually accepts indices as input.

Reply:
Oh nvm I wasn't using a `# type: ignore` for this function. Sorry about the confusion.

Reply:
I think the issue is entirely because you're using one class to implement part of the functionality of the other class, in a way that isn't how the API was designed.  It doesn't look like you're proposing an external API change - there's nothing really to break apart with `Token -> Index` and postprocessing, because all of that happens internal to the `TokenIndexer` in what you're doing here.

Reply:
To be clear, I'm not necessarily complaining about this usage of the API, as long as you have good tests, and comments explaining whatever type wrangling you have to do to make this work.  I was just saying that this looks to me like an internal implementation decision for how to handle long sequences, instead of something that requires rethinking of our external API.  If you think there is actually something larger here that should be external facing, we can talk about that.